packages/js-runtime/src/runtime/global/webscoket.ts

```diff
@@ -0,0 +1,241 @@
+      }
+
+      if (wsURL.hash !== '' || wsURL.href.endsWith('#')) {
+        throw new DOMException('Fragments are not allowed in a WebSocket URL.', 'SyntaxError');
+      }
+
+      this._url = wsURL.href;
+
+      let __protocols: string[];
+
+      if (typeof protocols === 'string') {
+        __protocols = [protocols];
+      } else {
+        __protocols = protocols;
+      }
+
+      if (protocols.length !== new Set([...__protocols.map(item => item.toLocaleLowerCase())]).size) {
+        throw new DOMException("Can't supply multiple times the same protocol.", 'SyntaxError');
+      }
+
+      LagonAsync.createWebsocket(wsURL.href, __protocols.join(', '))
+        .then(({ wsId, protocols, extensions }) => {
+          this.wsId = wsId;
+          this._protocols = protocols;
+          this._extensions = extensions;
+
+          if (this._readyState === WsState.CLOSING) {
+            LagonAsync.websocketClose(this.wsId).then(() => {
+              this._readyState = WsState.CLOSED;
+              const errEvent = new Event('error');
+              this.dispatchEvent(errEvent);
+              const event = new Event('close');
+              this.dispatchEvent(event);
+            });
+          } else {
+            this._readyState = WsState.OPEN;
+            const event = new Event('open');
+            this.dispatchEvent(event);
+
+            this.eventLoop();
+          }
+        })
+        .catch(err => {
+          this._readyState = WsState.CLOSED;
+          // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+          // @ts-ignore
+          const errorEv = new Event('error', { error: err, message: err.toString() });
+          this.dispatchEvent(errorEv);
+
+          const closeEv = new Event('close');
+          this.dispatchEvent(closeEv);
+        });
+    }
+
+    send(data: string | ArrayBuffer) {
+      if (this._readyState !== WsState.OPEN) {
+        throw new DOMException('readyState not OPEN', 'InvalidStateError');
+      }
+
+      LagonAsync.websocketSend(this.wsId, data);
+    }
+
+    close(code?: number, reason?: string) {
+      if (this._readyState === WsState.CONNECTING) {
+        this._readyState = WsState.CLOSING;
+      } else if (this._readyState === WsState.OPEN) {
+        this._readyState = WsState.CLOSING;
+
+        LagonAsync.websocketClose(this.wsId, code, reason).catch(err => {
+          const errorEv = new Event('error', {
+            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+            // @ts-ignore
+            error: err,
+            message: err.toString(),
+          });
+          this.dispatchEvent(errorEv);
+
+          const closeEv = new Event('close');
+          this.dispatchEvent(closeEv);
+        });
+      }
+    }
+
+    private async eventLoop() {
+      while (this._readyState !== WsState.CLOSED) {
+        try {
+          const data = await LagonAsync.websocketEvent(this.wsId);
```

Comment:
I don't agree with that. I believe this code involves modifying the state of the ws object in JavaScript and involves significant data interaction, such as creating an Event object. Therefore, I suggest performing the event loop on the JavaScript side.

Reply:
See my answer here: https://github.com/lagonapp/lagon/pull/887#discussion_r1201646254

Reply:
I reused the event loop in JS-land here, and I think it's not correct to handle events on the Native-land side. Firstly, as you can see from the code I wrote, I don't need much context information. I only need Native-land to tell me the current return value of ws, and then I trigger the corresponding subscriptions based on the return value. If I handle events on the Native-land side, the required content will not change, and triggering subscriptions can change the status of ws in JS-land, which can bring more costs. Maybe I have misunderstood your needs. If you have better ideas, can you directly modify this branch so that we can discuss further?